cxx/gendb.hh

```diff
@@ -0,0 +1,65 @@
+// Copyright 2024
+// See LICENSE.txt
+
+#pragma once
+#include <map>
+#include <random>
+#include <string>
+
+#include "distributions/crp.hh"
+#include "hirm.hh"
+#include "observations.hh"
+#include "pclean/schema.hh"
+#include "pclean/schema_helper.hh"
+
+class GenDB {
+ public:
+  const PCleanSchema& schema;
+
+  // TODO(emilyaf): Merge PCleanSchemaHelper and GenDB.
+  PCleanSchemaHelper schema_helper;
+
+  // This data structure contains entity sets and linkages. Semantics are
+  // map<class_name, map<primary_key, map<reference_field_name, ref_val>>>,
+  // where primary_key and ref_val are (integer) entity IDs.
+  std::map<std::string, std::map<int, std::map<std::string, int>>>
```

Comment:
In the current design, I don't think you ever need to say look up all of the primary keys given a class_name or anything like that.  So this could be replaced with a

std::map<std::tuple<std::string, std::string, int>, int>

and probably should, because that will be both more efficient (only one map lookup instead of three) and more informative to the reader.

It might also make sense to call this variable something like "foreign_keys", which foreign_keys[make_tuple(table_name, field_name, row_id)] giving you the foreign key that is entered into the row_id-th row of table_name in the field_name column.  (Which equals to row number in the table of field_name's type.)

Reply:
Good point, I changed the structure as you suggested.

I agree we should be better about unifying terminology. This PR refers to "reference" in a lot of places to essentially mean foreign keys (which the schema refers to as class variables). I'd rather stick with "reference_values" for now instead of introducing the "foreign keys" term and maybe change it later as part of a bigger renaming.